feat: mention local modules (#3877)
facebook/react-native-website: docs/local-library-setup.md

---
id: local-library-setup
title: Local libraries setup
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

> The local library is created outside of the `android/` and `ios/` folders and makes use of autolinking to integrate with your app. To create local library we will use [create-react-native-library](https://callstack.github.io/react-native-builder-bob/create). This tool contains all the necessary templates.

Reviewer: "The local library is created outside of..."

That's not the best way to start this page. Could we explain what are "local libraries" first rather than explaining the 'how' first?

Author: https://github.com/facebook/react-native-website/pull/3877/commits/642343b6bfee156b8cabe3d1ea3c7e4c9b788960

> The local library is created outside of the `android/` and `ios/` folders and makes use of autolinking to integrate with your app. To create local library we will use [create-react-native-library](https://callstack.github.io/react-native-builder-bob/create). This tool contains all the necessary templates.

Reviewer 2: ```suggestion
A local library is a library containing views or modules that's local to your app and not published to a registry. This is different from the traditional setup for view and modules in the sense that a local library is decoupled from your app's native code.

The local library is created outside of the `android/` and `ios/` folders and makes use of autolinking to integrate with your app. The structure with a local library may look like this:

\```
MyApp
├── node_modules
├── modules <-- folder for your local libraries
│   └── awesome-module <-- your local library
├── android
├── ios
├── src
├── index.js
└── package.json
\```

Since a local library's code exists outside of `android/` and `ios/` folders, it makes it easier to upgrade React Native versions in the future, copy to other projects etc.

To create local library we will use [create-react-native-library](https://callstack.github.io/react-native-builder-bob/create). This tool contains all the necessary templates.
```

### Getting Started

Inside your React Native application's root folder, run the following command:

```shell
npx create-react-native-library@latest awesome-module
```

Where `awesome-module` is the name you would like for the new module. After going through the prompts, you will have a new folder called `modules` in your project's root directory which contains the new module.

### Linking

By default, the generated library is automatically linked to the project using `link:` protocol when using Yarn and `file:` when using npm:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>

<TabItem value="npm">

```json
"dependencies": {
  "awesome-module": "file:./modules/awesome-module"
}
```

</TabItem>
<TabItem value="yarn">

```json
"dependencies": {
  "awesome-module": "link:./modules/awesome-module"
}
```

</TabItem>
</Tabs>

This creates a symlink to the library under `node_modules` which makes autolinking work.

### Installing dependencies

:::note

> Depending on module type you may have to make additional steps to make it work. Follow the instructions provided by the tool.

Reviewer 2: Need more clarification on what is meant by module type

Author: ```suggestion
Depending on whether you're creating a Native UI Component for custom visual elements or a Native Module to add custom native code for better performance or any other use case additional steps may be needed to make it work. Follow the instructions provided by the tool.
```

Reviewer 2: Hmm is there different instructions based on that? iirc it's the same setup for auto linking both views and modules.

Author: Actually not, let's remove this note callout.

> :::note
> Depending on module type you may have to make additional steps to make it work. Follow the instructions provided by the tool.
> :::
> 

Author: ```suggestion
```

To link the module you need to install dependencies:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>

<TabItem value="npm">

```shell
npm install

> ```

Reviewer 3: Should we make a call and only use `yarn`, @cortinico?
I remember having discussions that npm is brittle, but I'm, not totally sure.

Reviewer: In theory this should be tabbed npm/yarn as we do for all the other pages (like here https://reactnative.dev/docs/react-devtools)

</TabItem>
<TabItem value="yarn">

```shell
yarn install
```

</TabItem>
</Tabs>

### Using module inside your app

> To use the module inside your app, you need to import it from the `node_modules` folder:

Reviewer 2: ```suggestion
To use the module inside your app, you can import it by its name:
```
